frontend/src/metabase/common/components/DataPicker/components/DataPickerModal.tsx

```diff
@@ -122,99 +129,120 @@ export const DataPickerModal = ({
     [onChange, onClose, tryLogRecentItem],
   );
 
-  const handleCardChange = useCallback(
-    (item: QuestionPickerItem) => {
-      if (!isValidValueItem(item.model)) {
-        return;
-      }
-
-      onChange(getQuestionVirtualTableId(item.id));
-      tryLogRecentItem(item);
-      onClose();
-    },
-    [onChange, onClose, tryLogRecentItem],
-  );
-
-  const tabs: EntityTab<NotebookDataPickerValueItem["model"]>[] = [
-    hasModels && hasNestedQueriesEnabled
-      ? {
-          displayName: t`Models`,
-          model: "dataset" as const,
-          icon: "model",
-          element: (
-            <QuestionPicker
-              initialValue={isModelItem(value) ? value : undefined}
-              models={MODEL_PICKER_MODELS}
-              options={options}
-              shouldShowItem={modelsShouldShowItem}
-              onItemSelect={handleCardChange}
-            />
-          ),
-        }
-      : undefined,
-    hasMetrics && hasNestedQueriesEnabled
-      ? {
-          displayName: t`Metrics`,
-          model: "metric" as const,
-          icon: "metric",
-          element: (
-            <QuestionPicker
-              initialValue={isMetricItem(value) ? value : undefined}
-              models={METRIC_PICKER_MODELS}
-              options={options}
-              shouldShowItem={metricsShouldShowItem}
-              onItemSelect={handleCardChange}
-            />
-          ),
-        }
-      : undefined,
-    {
-      displayName: t`Tables`,
-      model: "table" as const,
-      icon: "table",
-      element: (
-        <TablePicker
-          databaseId={databaseId}
-          value={isTableItem(value) ? value : undefined}
-          onChange={handleChange}
-        />
-      ),
-    },
-    hasQuestions && hasNestedQueriesEnabled
-      ? {
-          displayName: t`Saved questions`,
-          model: "card" as const,
-          icon: "folder",
-          element: (
-            <QuestionPicker
-              initialValue={isQuestionItem(value) ? value : undefined}
-              models={QUESTION_PICKER_MODELS}
-              options={options}
-              shouldShowItem={questionsShouldShowItem}
-              onItemSelect={handleCardChange}
-            />
-          ),
-        }
-      : undefined,
-  ].filter(
-    (tab): tab is EntityTab<NotebookDataPickerValueItem["model"]> =>
-      tab != null && models.includes(tab.model),
-  );
+  const [modelsPath, setModelsPath] = useState<QuestionPickerStatePath>();
+  const [metricsPath, setMetricsPath] = useState<QuestionPickerStatePath>();
+  const [questionsPath, setQuestionsPath] = useState<QuestionPickerStatePath>();
+  const [tablesPath, setTablesPath] = useState<TablePickerStatePath>();
+
+  const tabs = (function getTabs() {
```

Comment:
nit: I don't use IIFE usually as function will need to be re-initialized on every rendering

Reply:
> function will need to be re-initialized on every rendering

This is negligible, it won't affect performance of the app.

FYI we talked about this during the offsite and concluded that IIFEs are actually a good way out.
Initially `tabs` were computed inside `useMemo` but there were many dependencies and it basically created a new array on every render, so `useMemo` actually degraded performance here. The only viable reason for the `useMemo` to be used here was that it grouped this code in a separate block. This is what IIFE does but without the overhead of deps comparison done inside the `useMemo`. Extracting a separate function out of this component is a bit awkward due to the number of arguments that function would need to receive (although it's preferred when you have very few arguments).

Let's keep it as is :pray: